File: Haskell/bouncingBall.hs
-- Define Ball data type with its center (x, y) and movement offsets (dx, dy).
data Ball = Ball { x :: Int
                 , y :: Int
                 , dx :: Int
                 , dy :: Int
                 } deriving(Show)

-- Define the arena dimensions.
arenaW = 480
arenaH = 360

-- Define the ball dimensions.
ballW = 20
ballH = 20

-- Moves the ball by its movement offset.
-- Params: a Ball.
-- Returns: the updated Ball, with its new position.
move :: Ball -> Ball
move (Ball x y dx dy)
    | x + dx > arenaW - ballW = move (Ball x y (-dx) dy)
    | x + dx < 0 = move (Ball x y (-dx) dy)
    | y + dy > arenaH - ballH = move (Ball x y dx (-dy))
    | y + dy < 0 = move (Ball x y dx (-dy))
    | otherwise = Ball (x + dx) (y + dy) dx dy

File: Rust/bouncing_ball.rs
struct Ball {
    x: i32,
    y: i32,
    dx: i32,
    dy: i32,
}

fn move_ball(ball: &mut Ball) {
    ball.x += ball.dx;
    ball.y += ball.dy;
}

fn main() {
    let mut ball = Ball {
        x: 0,
        y: 0,
        dx: 5,
        dy: 5,
    };

    move_ball(&mut ball);

    println!("x: {} y: {}", ball.x, ball.y);
}